RULES.md

```diff
@@ -0,0 +1,43 @@
+
+
+Ben's rules for writing performant database applications
+
+These are lessons that I have learned over a decade of writing Django and other
+Python web applications dealing with large amounts of data.  They might be
+unusual and controversial, so that's why I am wrote them down here.  They are
+not rules that we must follow in this project.  They are a guide for if and
+when we need to scale to large amounts of data or have fast database
+operations.
+
+1. Think in sets and do bulk operations.  A single entry is a set of one.   The
+   cost of a single insert or update is similar to the cost of a bulk insert or
+   update since most of the cost of database writes are updating indicies.
+
+2. Know what SQL statement that you want to send to the database first.  Use
+   tools to generate that SQL second. ORMs generate really slow SQL unless you
+   are very careful and know what you are doing.
+
+3. Don't override the save function.  This prevents #1.  Don't use most of ORM
```

Comment:
We don't have `save` function in SQLAlchemy. It's doesn't follow the Active Record pattern.